proposals/0002.md

```diff
@@ -0,0 +1,272 @@
+anyhow::Chain::new(&error).for_each(|e| print!(": {}", e);
+println!();
+```
+
+An example output:
+
+```
+An error occured: failed to load the configuration file: failed to load the file '~/.app_config': file not found
+```
+
+# Recommendation
+
+These examples of building errors serve as a practical guide.
+There are some common patterns of error structures and all of them can be implemented conveniently
+and correctly using `thiserror` and `anyhow`.
+
+All the examples use `enum`s, but it's fine to use `struct`s instead when there's only one variant.
+Usually it's not very useful, because it reduces flexibility, but doesn't simplify code much.
+
+## Root cause errors
+
+These are errors that don't wrap other errors inside.
+They may contain multiple structures and a lot of details, but their `source` always returns `None`.
+A root error can be just a variant of a bigger error with other variants, some of which may be even
+wrappers.
+
+### Statically typed root cause error
+
+The error has a statically defined meaning and details.
+It's easy to track it by humans and analyze it in the code.
+It requires scrutiny and planning to design, but it's easy to reuse in different parts of the code.
+
+```rust
+#[derive(Debug, thiserror::Error)]
+pub enum FileLoadError {
+    ...
+    #[error("the file contains invalid value {0}")
+    InvalidValue(u32),
+    ...
+}
+```
+
+### Textual root cause error
+
+The error is defined only as a text.
+It may be harder to track by humans it and it's almost impossible to analyze in the code.
+It's very easy to create new error variants, but they aren't easily reusable.
+
+Textual errors are often implemented as `String`s, but it's a subpar solution.
+It's much better to use `anyhow::Error`, because unlike `String` it implements
+`std::error::Error` making its processing easier and more consistent.
+The `anyhow` crate also provides the `anyhow!` macro, which works just like `format!`.
+
+```rust
+fn check_value(value: u32) -> Result<(), anyhow::Error> {
+    ...
+    .ok_or_else(|| anyhow::anyhow!("the file contains an unexpected value {}", value))?;
+    ...
+}
+```
+
+## Wrapper errors
+
+These are errors that wrap other errors inside. Their `source` always returns the inner error.
+A wrapper error can be just a variant of a bigger error with other variants, some of which may be
+even roots.
+
+### Statically typed wrapper error
+
+The error has a statically defined cause and details.
+It's easy to track it by humans and analyze it in the code.
+It requires scrutiny and planning to design, but it's easy to reuse in different parts of the code.
+
+```rust
+#[derive(Debug, thiserror::Error)]
+pub enum FileLoadError {
+    ...
+    #[error("failed to open the file '{path}'")]
+    FileSystemError {
+        path: String,
+        #[source]
+        source: std::io::Error,
+    },
+    ...
+}
+```
+
+Sometimes it's necessary to wrap an error not as a source, but as another case of the wrapper.
```

Comment:
I found the original sentence a bit confusing as to what you were describing, but the example made sense. Maybe you would prefer this wording or something like it :)

Suggested wording:
Sometimes we would like a variant to act like the error it contains. We delegate the `Error` implementation to the underlying error.

Reply:
Fixed with the help of your suggestion. Is it good now?